pkg/util/log/sampled.go

```diff
@@ -0,0 +1,41 @@
+package log
+
+import (
+	"context"
+	"time"
+
+	"github.com/weaveworks/common/logging"
+	"github.com/weaveworks/common/user"
+	"go.uber.org/atomic"
+)
+
+type SampledError struct {
+	Err     error
+	Sampler *Sampler
+}
+
+func (s SampledError) Error() string { return s.Err.Error() }
+func (s SampledError) Unwrap() error { return s.Err }
+
+// This method is called by grpc_logging.go, instead of doing its own logging.
+func (s SampledError) Observe(ctx context.Context, log logging.Interface, method string, duration time.Duration) {
+	if !s.Sampler.Sample() {
+		return // no-op
+	}
+	entry := user.LogWith(ctx, log).WithField("method", method).WithField("duration", duration)
+	entry.WithField("err", s.Err).Warnln("gRPC")
+}
+
+type Sampler struct {
+	freq  int
+	count atomic.Int64
+}
+
+func NewSampler(freq int) *Sampler {
+	return &Sampler{freq: freq}
+}
+
+func (e *Sampler) Sample() bool {
+	count := e.count.Inc()
+	return count%int64(e.freq) == 0
```

Comment:
wdyt of making `freq` an int64 so that you don't have to alloc it on `Sample()` invocations?

Reply:
I don't think there is an allocation, but sure, the cast is unnecessary if I change the type.

Reply:
hah I thought the cast would end up allocating. ty